server/types/bool_array.go

```diff
@@ -0,0 +1,265 @@
+// Copyright 2024 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package types
+
+import (
+	"bytes"
+	"math"
+	"reflect"
+	"strings"
+
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/types"
+	"github.com/dolthub/vitess/go/sqltypes"
+	"github.com/dolthub/vitess/go/vt/proto/query"
+
+	"github.com/dolthub/doltgresql/utils"
+)
+
+// BoolArray is the standard boolean array.
+var BoolArray = BoolArrayType{}
+
+// BoolArrayType is the extended type implementation of the PostgreSQL boolean.
+type BoolArrayType struct{}
+
+var _ types.ExtendedType = BoolArrayType{}
+
+// CollationCoercibility implements the types.ExtendedType interface.
+func (b BoolArrayType) CollationCoercibility(ctx *sql.Context) (collation sql.CollationID, coercibility byte) {
+	return sql.Collation_binary, 5
+}
+
+// Compare implements the types.ExtendedType interface.
+func (b BoolArrayType) Compare(v1 any, v2 any) (int, error) {
+	if v1 == nil && v2 == nil {
+		return 0, nil
+	} else if v1 != nil && v2 == nil {
+		return 1, nil
+	} else if v1 == nil && v2 != nil {
+		return -1, nil
+	}
+
+	ac, _, err := b.Convert(v1)
+	if err != nil {
+		return 0, err
+	}
+	bc, _, err := b.Convert(v2)
+	if err != nil {
+		return 0, err
+	}
+
+	ab := ac.([]bool)
+	bb := bc.([]bool)
+	minLength := utils.Min(len(ab), len(bb))
+	for i := 0; i < minLength; i++ {
+		if ab[i] == bb[i] {
+			continue
+		} else if !ab[i] {
+			return -1, nil
+		} else {
+			return 1, nil
+		}
+	}
+	if len(ab) == len(bb) {
+		return 0, nil
+	} else if len(ab) < len(bb) {
+		return -1, nil
+	} else {
+		return 1, nil
+	}
+}
+
+// Convert implements the types.ExtendedType interface.
+func (b BoolArrayType) Convert(val any) (any, sql.ConvertInRange, error) {
+	if val == nil {
+		return nil, sql.InRange, nil
+	}
+
+	switch val := val.(type) {
```

Comment:
Probably need to figure out the semantics here, presumably other conversions are possible but I don't know how they work